source/windows/bcrypt_rsa.c

```diff
@@ -0,0 +1,459 @@
+        padding_info->pszAlgId = BCRYPT_SHA256_ALGORITHM;
+        padding_info->cbSalt = 32;
+        return padding_info;
+    }
+
+    return NULL;
+}
+
+int s_rsa_sign(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signing_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out) {
+    struct bcrypt_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    void *padding_info = s_create_sign_padding_info(key_pair->allocator, algorithm);
+    if (padding_info == NULL) {
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    size_t output_buf_space = out->capacity - out->len;
+
+    if (output_buf_space < aws_rsa_key_pair_signature_length(key_pair)) {
+        return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+    }
+
+    struct aws_byte_buf temp_signature_buf;
+    aws_byte_buf_init(&temp_signature_buf, key_pair->allocator, aws_rsa_key_pair_signature_length(key_pair));
+    size_t signature_length = temp_signature_buf.capacity;
+
+    NTSTATUS status = BCryptSignHash(
+        key_pair_impl->key_handle,
+        padding_info,
+        digest.ptr,
+        (ULONG)digest.len,
+        temp_signature_buf.buffer,
+        (ULONG)signature_length,
+        (ULONG *)&signature_length,
+        algorithm == AWS_CAL_RSA_SIGNATURE_PKCS1_5_SHA256 ? BCRYPT_PAD_PKCS1 : BCRYPT_PAD_PSS);
+
+    if (!BCRYPT_SUCCESS(status)) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        goto on_error;
+    }
+
+    temp_signature_buf.len = signature_length;
+
+    struct aws_byte_cursor temp_signature_cur = aws_byte_cursor_from_buf(&temp_signature_buf);
+    aws_byte_buf_append(out, &temp_signature_cur);
+
+    aws_byte_buf_clean_up_secure(&temp_signature_buf);
+    aws_mem_release(key_pair->allocator, padding_info);
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+    aws_byte_buf_clean_up_secure(&temp_signature_buf);
+    aws_mem_release(key_pair->allocator, padding_info);
+    return AWS_OP_ERR;
+}
+
+int s_rsa_verify(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signing_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature) {
+    struct bcrypt_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    void *padding_info = s_create_sign_padding_info(key_pair->allocator, algorithm);
+    if (padding_info == NULL) {
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    /* okay, now we've got a windows compatible signature, let's verify it. */
+    NTSTATUS status = BCryptVerifySignature(
+        key_pair_impl->key_handle,
+        padding_info,
+        digest.ptr,
+        (ULONG)digest.len,
+        signature.ptr,
+        (ULONG)signature.len,
+        algorithm == AWS_CAL_RSA_SIGNATURE_PKCS1_5_SHA256 ? BCRYPT_PAD_PKCS1 : BCRYPT_PAD_PSS);
+
+    aws_mem_release(key_pair->allocator, padding_info);
+    return status == 0 ? AWS_OP_SUCCESS : aws_raise_error(AWS_ERROR_CAL_SIGNATURE_VALIDATION_FAILED);
```

Comment:
are we sure that's why it failed? is there a status code we need to check?

Reply:
--> STATUS_INVALID_SIGNATURE <--